src/profiles_management/pmr/classes.py

```diff
@@ -0,0 +1,133 @@
+"""This package provides classes for validating and defining a PMR.
+
+The goal of these classes is to represent the PMR in a more
+Pythonic way, rather than a dict of vavlues.
+
+The classes should also provide some common helper methods for
+traversing the information of the PMR, but not functions for
+directly affecting the K8s cluster based on the PMR.
+
+This is similar to how the K8s python clients have classes
+for representing the objects and different methods for doing operations
+with them.
+"""
+
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, List
+
+import jsonschema
+
+from profiles_management.pmr.schema import RESOURCE_QUOTA_SCHEMA
+
+log = logging.getLogger(__name__)
+
+
+class UserKind(Enum):
+    """Class representing the kind of the user."""
+
+    USER = "user"
+    SERVICE_ACCOUNT = "service-account"
+
+
+class ContributorRole(Enum):
+    """Class representing the role of the user."""
+
+    ADMIN = "admin"
+    EDIT = "edit"
+    VIEW = "view"
+
+
+@dataclass
+class Contributor:
+    """Class representing which users should have access to a Profile."""
+
+    name: str
+    role: ContributorRole
+
+
+@dataclass
+class Owner:
+    """Class representing the owner field of a Profile."""
+
+    name: str
+    kind: UserKind
+
+
+@dataclass
+class Profile:
+    """Class representing a Profile and its Contributors."""
+
+    name: str
+    owner: Owner
+    resources: Dict[str, Any]
+    contributors: List[Contributor]
+
+    # https://docs.python.org/3/library/dataclasses.html#post-init-processing
+    def __post_init__(self):
+        """Validate resourceQuota after values have been initialised."""
+        log.info("Validating ResourceQuota for Profile: %s", self.name)
+        jsonschema.validate(self.resources, RESOURCE_QUOTA_SCHEMA)
+        log.info("ResourceQuota is valid.")
+
+    def has_contributor(self, name: str, role: ContributorRole) -> bool:
+        """Check if the Profile has a contributor with specific role."""
+        for contributor in self.contributors:
+            if contributor.name != name:
+                continue
+
+            if contributor.role != role:
+                continue
+
+            log.info("Profile %s has contributor (%s, %s)", self.name, name, role.value)
+            return True
+
+        log.info("Profile %s doesn't have contributor (%s, %s)", self.name, name, role.value)
+        return False
+
+
+class ProfilesManagementRepresentation:
+    """A class representing the Profiles and Contributors."""
+
+    profiles: dict[str, Profile] = {}
```

Comment:
We also have the following issue:

```python
p1 = classes.Profile(name='daniela-p1', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
p2 = classes.Profile(name='daniela-p2', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
p3 = classes.Profile(name='daniela-p3', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
pmr1 = classes.ProfilesManagementRepresentation()
pmr1.add_profile(p1)
pmr1.add_profile(p2)
pmr1.add_profile(p3)

pmr2 = classes.ProfilesManagementRepresentation([p1])

# Print pmr1
Profiles:
-  daniela-p1: (daniela, edit)
-  daniela-p2: (daniela, edit)
-  daniela-p3: (daniela, edit)

# Print pmr2
Profiles:
-  daniela-p1: (daniela, edit)
-  daniela-p2: (daniela, edit)
-  daniela-p3: (daniela, edit)
```

Reply:
I'm not sure if the above is a problem or if we try to create problems 😅 

What would you suggest that we do instead?

Reply:
Pulling up the terminal

Reply:
OK, got in sync with the first comment, nice catch! Yep, the error is because the `profiles` attribute is not initialised during `__init__`.

Updated the code
6dc2b70

Reply:
Mmmh, so there is no reason to have `profiles` as a class attribute? If not, we can do the following to avoid it being so accessible outside of the class:

```diff
diff --git a/src/profiles_management/pmr/classes.py b/src/profiles_management/pmr/classes.py
index 99ab8cd..64e1bec 100644
--- a/src/profiles_management/pmr/classes.py
+++ b/src/profiles_management/pmr/classes.py
@@ -123,9 +123,7 @@ class ProfilesManagementRepresentation:
         ValidationError: From pydantic if the validation failed.
     """

-    profiles: dict[str, Profile] = {}
-
-    def __init__(self, profiles_list: List[Profile] = []):
+    def __init__(self, profiles_list: List[Profile] = None):
         """Initialise based on a list of Profiles.

         If a list of Profiles is given, then the internal dict will be initialised
@@ -138,11 +136,20 @@ class ProfilesManagementRepresentation:
             ValidationError: From pydantic if the validation failed.
         """
         # validate the input type
-        TypeAdapter(List[Profile]).validate_python(profiles_list)
-
-        self.profiles = {}
-        for profile in profiles_list:
-            self.add_profile(profile)
+        self._profiles = None
+        self._profiles_list = profiles_list
+        if profiles_list: # <-- if you have a better way of doing this validation, feel free to change it
+          TypeAdapter(List[Profile]).validate_python(profiles_list)
+
+
+    @property
+    def profiles(self):
+      if not self._profiles:
+        profiles_dict = {}
+        for profile in self._profiles_list or []:
+          profiles_dict[profile.name] = profile
+        self._profiles = profiles_dict
+      return self._profiles

     def has_profile(self, name: str) -> bool:
         """Check if given Profile name is part of the PMR.
```

With this we'll avoid both issues from above:

```
# cat test_pmr_classes.py

from profiles_management.pmr import classes
p1 = classes.Profile(name='daniela-p1', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
p2 = classes.Profile(name='daniela-p2', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
p3 = classes.Profile(name='daniela-p3', owner=classes.Owner(name="daniela", kind=classes.UserKind.USER), resources={}, contributors=[classes.Contributor(name='daniela', role=classes.ContributorRole.EDIT)])
pmr1 = classes.ProfilesManagementRepresentation()
pmr1.add_profile(p1)
pmr1.add_profile(p2)
pmr1.add_profile(p3)
print(pmr1)
pmr1.remove_profile(p3.name)
print(pmr1)

pmr2 = classes.ProfilesManagementRepresentation([p1])
print(pmr2)

pmr3 = classes.ProfilesManagementRepresentation()
print(pmr3)

# Outputs
$ python3 test_pmr_classes.py
Profiles:
-  daniela-p1: (daniela, edit)
-  daniela-p2: (daniela, edit)
-  daniela-p3: (daniela, edit)

Profiles:
-  daniela-p1: (daniela, edit)
-  daniela-p2: (daniela, edit)

Profiles:
-  daniela-p1: (daniela, edit)

Profiles:
```

Reply:
Small note that we can't make `profiles_list: List[Profile] = None` as there's this warning
```
Expression of type "None" cannot be assigned to parameter of type "List[Profile]"
  "None" is not assignable to "List[Profile]" [reportArgumentType]
```

Could you explain me a bit more what we would gain from using a `@property` here? 

With the current commit I pushed https://github.com/canonical/github-profiles-automator/commit/6dc2b70faeb0f87a623f7a24e58609a2ae361d90, that makes `profiles` an instance attribute and not a class attribute, I see the exact same behavior as the one you've put above.

Reply:
> Small note that we can't make profiles_list: List[Profile] = None as there's this warning

In my diff I worked around that, please check it.

> Could you explain me a bit more what we would gain from using a @property here?

We won't have a class attribute that can be overwritten by anything else.

> that makes profiles an instance attribute and not a class attribute, I see the exact same behavior as the one you've put above.

are you sure about this? I was still able to access and overwrite its values from any member of the class, and even from outside the class.

Reply:
The snippet you showed throws the lint error I showed
![image](https://github.com/user-attachments/assets/2d88f2b3-f085-4c72-8c37-8b513e81ad96)

Since my commit https://github.com/canonical/github-profiles-automator/commit/6dc2b70faeb0f87a623f7a24e58609a2ae361d90 already addresses the class / instance attribute issue with one line of change, can we please not block the PR any longer? cc @deusebio 

The benefit of the approach you showed will not allow other places in the code to do `pmr.profiles = ...` since there's no setter. But we can live without this, since it adds lint error and more complex logic for a simple attribute.

Reply:
I've kept the input as empty list, as I couldn't manage to make it work with `None` and `ruff` to not complain. Let me know if you have other suggestions.

I also removed the conditional validation for `profiles_list` and had it run every time, as IIUC we want to run the validation always. Plus users could omit the value when calling the constructor and the default value `[]` will be used.

One last thing that I'd like to confirm is what happens when `add_profile(profile)` is being called:
1. The code does `self.profiles[profile.name] = profile`
2. The `profiles` property function gets triggered, which ends up returning `self._profiles` dict
3. This is the dict then in which the new element gets added

This is your understanding right?

I also tried it and can confirm that the other helper methods of the PMR are working and the tests are passing.
c598d9e